In [57]:
import pandas as pd

<h3> Let's analyze climate dataset </h3> 


In [58]:
df_climate = pd.read_csv('climate.csv')

<h3> Look for the column data type. This might be important to know numeric vs string columns </h3> 

In [59]:
df_climate.dtypes

lng                         float64
lat                         float64
STATION_NAME                 object
CLIMATE_IDENTIFIER           object
ID                           object
LOCAL_DATE                   object
PROVINCE_CODE                object
LOCAL_YEAR                    int64
LOCAL_MONTH                   int64
LOCAL_DAY                     int64
MEAN_TEMPERATURE            float64
MEAN_TEMPERATURE_FLAG        object
MIN_TEMPERATURE             float64
MIN_TEMPERATURE_FLAG         object
MAX_TEMPERATURE             float64
MAX_TEMPERATURE_FLAG         object
TOTAL_PRECIPITATION         float64
TOTAL_PRECIPITATION_FLAG     object
TOTAL_RAIN                  float64
TOTAL_RAIN_FLAG              object
TOTAL_SNOW                  float64
TOTAL_SNOW_FLAG              object
SNOW_ON_GROUND              float64
SNOW_ON_GROUND_FLAG          object
DIRECTION_MAX_GUST          float64
DIRECTION_MAX_GUST_FLAG      object
SPEED_MAX_GUST              float64
SPEED_MAX_GUST_FLAG         

<h3> Look for the dataframe stats </h3>

In [60]:
df_climate.describe()

lng           lat  LOCAL_YEAR   LOCAL_MONTH     LOCAL_DAY  \
count  41513.000000  41513.000000     41513.0  41513.000000  41513.000000   
mean     -98.869026     52.456319      2021.0      1.325585     13.395129   
std       23.178858      6.741840         0.0      0.468599      8.607924   
min     -140.868889     41.948614      2021.0      1.000000      1.000000   
25%     -117.817094     48.647222      2021.0      1.000000      6.000000   
50%     -105.137222     50.441111      2021.0      1.000000     12.000000   
75%      -78.166667     54.286111      2021.0      2.000000     20.000000   
max      -52.783339     76.425833      2021.0      2.000000     31.000000   

       MEAN_TEMPERATURE  MIN_TEMPERATURE  MAX_TEMPERATURE  \
count      39456.000000     39573.000000     39538.000000   
mean         -11.061200       -15.364056        -6.731327   
std           11.507346        12.235715        11.262898   
min          -50.000000       -51.800000       -48.100000   
25%          -19.200000       -24.900000       -13.900000   
50%           -8.700000       -13.600000        -3.700000   
75%           -2.300000        -5.700000         1.500000   
max           11.500000         9.800000        16.500000   

       TOTAL_PRECIPITATION   TOTAL_RAIN   TOTAL_SNOW  SNOW_ON_GROUND  \
count         30441.000000  7895.000000  7934.000000    21865.000000   
mean              2.250977     2.565763     1.499370       26.529842   
std               6.777407     8.358784     4.098045       26.587205   
min               0.000000     0.000000     0.000000        0.000000   
25%               0.000000     0.000000     0.000000        7.000000   
50%               0.000000     0.000000     0.000000       22.000000   
75%               1.400000     0.000000     1.000000       38.000000   
max             292.500000   163.500000    70.000000      258.000000   

       DIRECTION_MAX_GUST  SPEED_MAX_GUST  COOLING_DEGREE_DAYS  \
count        18403.000000    18585.000000              39456.0   
mean            23.130414       47.406726                  0.0   
std              9.459961       15.174350                  0.0   
min              1.000000       31.000000                  0.0   
25%             17.000000       36.000000                  0.0   
50%             26.000000       43.000000                  0.0   
75%             31.000000       54.000000                  0.0   
max             36.000000      191.000000                  0.0   

       HEATING_DEGREE_DAYS  MIN_REL_HUMIDITY  MAX_REL_HUMIDITY  
count         39456.000000      33531.000000      33531.000000  
mean             29.061200         67.547434         86.593719  
std              11.507346         12.859809          9.364097  
min               6.500000         14.000000         43.000000  
25%              20.300000         60.000000         80.000000  
50%              26.700000         68.000000         88.000000  
75%              37.200000         76.000000         94.000000  
max              68.000000        100.000000        100.000000

<h5><i> There are some missing measurements for all of the temperature related metrics. 
        Example: Total records = 41513, Records with non-null MIN_TEMPERATURE  = 39573
    </i></h5>

<h3> Look for the missing values </h3>

In [61]:
df_climate[df_climate.MEAN_TEMPERATURE.isnull()][['lng', 'lat', 'STATION_NAME', 'CLIMATE_IDENTIFIER', 'ID', 'LOCAL_DATE',
       'PROVINCE_CODE', 'LOCAL_YEAR', 'LOCAL_MONTH', 'LOCAL_DAY',
       'MEAN_TEMPERATURE', 'MEAN_TEMPERATURE_FLAG', 'MIN_TEMPERATURE',
       'MIN_TEMPERATURE_FLAG']].head(10)


lng        lat           STATION_NAME CLIMATE_IDENTIFIER  \
54  -112.279167  53.916944            ANDREW AGDM            3010237   
64  -112.279167  53.916944            ANDREW AGDM            3010237   
87  -112.279167  53.916944            ANDREW AGDM            3010237   
488 -113.466667  53.666667  EDMONTON NAMAO AWOS A            3012212   
516 -113.579444  53.310000        EDMONTON INTL A            3012216   
517 -113.579444  53.310000        EDMONTON INTL A            3012216   
644 -111.000000  52.966667                 FABYAN            3012515   
650 -111.000000  52.966667                 FABYAN            3012515   
656 -111.000000  52.966667                 FABYAN            3012515   
662 -111.000000  52.966667                 FABYAN            3012515   

                    ID       LOCAL_DATE PROVINCE_CODE  LOCAL_YEAR  \
54    3010237.2020.1.9  2021-01-09 0:00            AB        2021   
64   3010237.2020.1.19  2021-01-19 0:00            AB        2021   
87   3010237.2020.2.11  2021-02-11 0:00            AB        2021   
488  3012212.2020.1.29  2021-01-29 0:00            AB        2021   
516  3012216.2020.1.11  2021-01-11 0:00            AB        2021   
517  3012216.2020.1.12  2021-01-12 0:00            AB        2021   
644   3012515.2020.1.1  2021-01-01 0:00            AB        2021   
650  3012515.2020.1.16  2021-01-16 0:00            AB        2021   
656  3012515.2020.1.25  2021-01-25 0:00            AB        2021   
662   3012515.2020.2.3  2021-02-03 0:00            AB        2021   

     LOCAL_MONTH  LOCAL_DAY  MEAN_TEMPERATURE MEAN_TEMPERATURE_FLAG  \
54             1          9               NaN                     M   
64             1         19               NaN                     M   
87             2         11               NaN                     M   
488            1         29               NaN                     M   
516            1         11               NaN                     M   
517            1         12               NaN                     M   
644            1          1               NaN                     M   
650            1         16               NaN                     M   
656            1         25               NaN                     M   
662            2          3               NaN                     M   

     MIN_TEMPERATURE MIN_TEMPERATURE_FLAG  
54               NaN                    M  
64               NaN                    M  
87               NaN                    M  
488              NaN                    M  
516              NaN                    M  
517              NaN                    M  
644            -10.0                  NaN  
650            -40.0                  NaN  
656            -13.5                  NaN  
662            -21.5                  NaN

In [62]:
df_climate[(abs(df_climate.MEAN_TEMPERATURE) >= 0) & (df_climate.MEAN_TEMPERATURE_FLAG == 'M')][['lng', 'lat', 'STATION_NAME', 'CLIMATE_IDENTIFIER', 'ID', 'LOCAL_DATE',
       'PROVINCE_CODE', 'LOCAL_YEAR', 'LOCAL_MONTH', 'LOCAL_DAY',
       'MEAN_TEMPERATURE', 'MEAN_TEMPERATURE_FLAG', 'MIN_TEMPERATURE',
       'MIN_TEMPERATURE_FLAG']].head(5)


Empty DataFrame
Columns: [lng, lat, STATION_NAME, CLIMATE_IDENTIFIER, ID, LOCAL_DATE, PROVINCE_CODE, LOCAL_YEAR, LOCAL_MONTH, LOCAL_DAY, MEAN_TEMPERATURE, MEAN_TEMPERATURE_FLAG, MIN_TEMPERATURE, MIN_TEMPERATURE_FLAG]
Index: []

<h5><i> Apparently when MEAN_TEMPERATURE_FLAG == 'M', there is no reading for MEAN_TEMPERATURE for that day
    </i></h5>

In [63]:
df_climate[(df_climate.MEAN_TEMPERATURE.isnull()) & (df_climate.MEAN_TEMPERATURE_FLAG != 'M')][['lng', 'lat', 'STATION_NAME', 'CLIMATE_IDENTIFIER', 'ID', 'LOCAL_DATE',
       'PROVINCE_CODE', 'LOCAL_YEAR', 'LOCAL_MONTH', 'LOCAL_DAY',
       'MEAN_TEMPERATURE', 'MEAN_TEMPERATURE_FLAG', 'MIN_TEMPERATURE',
       'MIN_TEMPERATURE_FLAG']].head(5)


lng        lat STATION_NAME CLIMATE_IDENTIFIER                ID  \
1039 -110.55  52.016667      MONITOR            3014600  3014600.2020.1.1   
1040 -110.55  52.016667      MONITOR            3014600  3014600.2020.1.2   
1041 -110.55  52.016667      MONITOR            3014600  3014600.2020.1.3   
1042 -110.55  52.016667      MONITOR            3014600  3014600.2020.1.4   
1043 -110.55  52.016667      MONITOR            3014600  3014600.2020.1.5   

           LOCAL_DATE PROVINCE_CODE  LOCAL_YEAR  LOCAL_MONTH  LOCAL_DAY  \
1039  2021-01-01 0:00            AB        2021            1          1   
1040  2021-01-02 0:00            AB        2021            1          2   
1041  2021-01-03 0:00            AB        2021            1          3   
1042  2021-01-04 0:00            AB        2021            1          4   
1043  2021-01-05 0:00            AB        2021            1          5   

      MEAN_TEMPERATURE MEAN_TEMPERATURE_FLAG  MIN_TEMPERATURE  \
1039               NaN                   NaN              NaN   
1040               NaN                   NaN              NaN   
1041               NaN                   NaN              NaN   
1042               NaN                   NaN              NaN   
1043               NaN                   NaN              NaN   

     MIN_TEMPERATURE_FLAG  
1039                  NaN  
1040                  NaN  
1041                  NaN  
1042                  NaN  
1043                  NaN

<h5><i> Apparently when MEAN_TEMPERATURE_FLAG == 'M', there is no reading for that MEAN_TEMPERATURE for that day
    </i></h5>

In [64]:
df_climate[(df_climate.MEAN_TEMPERATURE.isnull()) & (df_climate.MEAN_TEMPERATURE_FLAG != 'M')][['lng', 'lat', 'STATION_NAME', 'CLIMATE_IDENTIFIER', 'ID', 'LOCAL_DATE',
       'PROVINCE_CODE', 'LOCAL_YEAR', 'LOCAL_MONTH', 'LOCAL_DAY',
       'MEAN_TEMPERATURE', 'MEAN_TEMPERATURE_FLAG', 'MIN_TEMPERATURE',
       'MIN_TEMPERATURE_FLAG']].STATION_NAME.unique()

array(['MONITOR', 'METCHOSIN', 'NORTH COWICHAN', 'SAANICHTON MT NEWTON',
       'NANAIMO CITY YARD', 'BURNABY SIMON FRASER U',
       'N VANCOUVER WHARVES', 'PORTAGE LA PRAIRIE CDA', 'BEAVER',
       'FREDERICTON AQUATIC CENTRE C/S', 'GAGETOWN A', 'ROCKY HARBOUR CS',
       'SAGLEK', 'WHITE ROCK', 'CAPE PARRY A', 'HAT ISLAND',
       'SHEPHERD BAY A', 'CAPE PEEL WEST', 'DEWAR LAKES', 'ROWLEY ISLAND',
       'BORDEN AWOS', "VAL-D'OR A", 'SHINGLE POINT A'], dtype=object)

<h5><i> Even if the flag is null, still the reading for that metric could be null. An investigation is required to understand why we are having no reading for some days.</i></h5>

<h3> Look for business key. Business key is an unique identifier of the station </h3>

In [65]:
df_climate['CLIMATE_IDENTIFIER'].describe()

count       41513
unique        958
top       4061632
freq           47
Name: CLIMATE_IDENTIFIER, dtype: object

<h5><i> Climate identifier is a non-null field as it counts = size of th dataframe. Also based on google search, I was able to conclude that lat and lng are the location of that weather station</i></h5>

<h3> Analyzing city dataset </h3>


In [66]:
df_city = pd.read_csv('cities.csv')

In [67]:
df_city.head(10)

city        lat         lng country iso2             admin  capital  \
0    Toronto  43.666667  -79.416667  Canada   CA           Ontario    admin   
1   Montréal  45.500000  -73.583333  Canada   CA            Québec      NaN   
2  Vancouver  49.250000 -123.133333  Canada   CA  British Columbia      NaN   
3     Ottawa  45.416667  -75.700000  Canada   CA           Ontario  primary   
4    Calgary  51.083333 -114.083333  Canada   CA           Alberta      NaN   
5   Edmonton  53.550000 -113.500000  Canada   CA           Alberta    admin   
6   Hamilton  43.256101  -79.857484  Canada   CA           Ontario      NaN   
7   Winnipeg  49.883333  -97.166667  Canada   CA          Manitoba    admin   
8     Québec  46.800000  -71.250000  Canada   CA            Québec    admin   
9     Oshawa  43.900000  -78.866667  Canada   CA           Ontario      NaN   

   population  population_proper  
0     5213000            3934421  
1     3678000            2356556  
2     2313328             603502  
3     1145000             812129  
4     1110000             915322  
5     1058000             712391  
6      721053             519949  
7      632063             575313  
8      624177             528595  
9      450963             247989

<h5><i> At first glance, For Québec and Montréal we have special character é. </i></h5>

<h3> Look for the column data type. This might be important to know numeric vs string columns </h3> 

In [68]:
df_city.dtypes

city                  object
lat                  float64
lng                  float64
country               object
iso2                  object
admin                 object
capital               object
population             int64
population_proper      int64
dtype: object

<h3> Look for the dataframe stats </h3>

In [69]:
df_city.describe()

lat         lng    population  population_proper
count  247.000000  247.000000  2.470000e+02       2.470000e+02
mean    52.922777  -91.917140  9.365829e+04       6.671848e+04
std      7.340857   22.935838  4.510522e+05       3.120743e+05
min     42.301649 -139.450000  0.000000e+00       0.000000e+00
25%     48.183334 -112.750000  1.089000e+03       1.066000e+03
50%     50.933333  -89.233333  5.043000e+03       4.331000e+03
75%     55.783333  -73.096227  2.591750e+04       2.435050e+04
max     82.483333  -52.666667  5.213000e+06       3.934421e+06

<h5><i> No null columns for numeric values. </i></h5>

In [70]:
df_city[['city', 'country', 'iso2', 'admin', 'capital']].describe()

city country iso2    admin capital
count       247     247  247      247      14
unique      245       1    1       13       2
top     Windsor  Canada   CA  Ontario   admin
freq          2     247  247       47      13

<h5><i> Null value just for capital column. admin represents province. </i></h5>

In [71]:
df_city['capital'].unique()

array(['admin', nan, 'primary'], dtype=object)

In [72]:
df_city[df_city.capital == 'admin'].head(10)

city        lat         lng country iso2  \
0         Toronto  43.666667  -79.416667  Canada   CA   
5        Edmonton  53.550000 -113.500000  Canada   CA   
7        Winnipeg  49.883333  -97.166667  Canada   CA   
8          Québec  46.800000  -71.250000  Canada   CA   
11        Halifax  44.650000  -63.600000  Canada   CA   
14       Victoria  48.450234 -123.343529  Canada   CA   
17         Regina  50.450000 -104.616667  Canada   CA   
22   Saint John’s  47.550000  -52.666667  Canada   CA   
39    Fredericton  45.910648  -66.658649  Canada   CA   
46  Charlottetown  46.238225  -63.139481  Canada   CA   

                        admin capital  population  population_proper  
0                     Ontario   admin     5213000            3934421  
5                     Alberta   admin     1058000             712391  
7                    Manitoba   admin      632063             575313  
8                      Québec   admin      624177             528595  
11                Nova Scotia   admin      359111             222874  
14           British Columbia   admin      289625             251358  
17               Saskatchewan   admin      176183             176183  
22  Newfoundland and Labrador   admin      131469              99182  
39              New Brunswick   admin       52337              36713  
46       Prince Edward Island   admin       42402              31293

<h5><i> Admin represents record which corresponds to capital city of the provinces . </i></h5>

In [73]:
df_city[df_city.capital == 'primary'].head(10)

city        lat   lng country iso2    admin  capital  population  \
3  Ottawa  45.416667 -75.7  Canada   CA  Ontario  primary     1145000   

   population_proper  
3             812129

<h5><i> primary represents record which corresponds to capital city of Canada . </i></h5>

<h3> Relate two datasets </h3>


-> Determine the city where the weather station is located <br/>
-> We have province as a common column between two datasets, so that narrows down our search for the city. <br/>
-> We can join the two datasets on province. <br/>
-> Then find the proximity of weather station to the cities of that province. <br/>
-> Calculate abs(city_lat) - abs(station_lat) , abs(city_lng) - abs(station_lng) <br/>
-> Get the abs average of the differences <br/>
-> Drop all the records for that station except for the one whose average distance is the lowest. <br/> 
(Here I am assuming we won't have same difference in proximity for 2 or more differnt cities) <br/>
